### DISCUSSION TITLE
What can clients do to best handle the API not showing mempool txs issue?

### DISCUSSION THREAD

**Discussion by kyranjamie** [blockstack/stacks-wallet-web#1727] [General]
Opening this to discuss what wallet clients can do to best handle the API issues described in https://github.com/hirosystems/devops/issues/797, where newly broadcast transactions don't appear in the mempool

cc/ @yknl, author of Xverse wallet

**Comment by markmhendrickson** [comment]
It seems like the key question here is whether we can improve nonce value setting on the client-side by tracking all recent transactions as submitted to the API on the client-side, in the scenario that the API doesn't return any information about these transactions at any point after submission.

I've been presuming with https://github.com/blockstack/stacks-wallet-web/issues/1715 that this would be helpful, since the client could keep upping the nonce consistently with the combination of client-side data (for new transactions placed in "submitted" state per [this discussion](https://github.com/blockstack/stacks-wallet-web/discussions/1712)) and server-side data (for previous transactions). @aulneau appears to have the same idea with https://github.com/blockstack/stacks-wallet-web/pull/1726.

Note that this would help only users of the same wallet installation, since obviously if they go to another one, the cached state there won't help with nonce calculations on the client-side. I presume, however, that most users affected by the API propagation issues are facing them only within the context of a single installation.

However, @yknl and @kyranjamie appear doubtful this will help and might even exacerbate issues with nonce setting. Perhaps we should spell out some particular scenarios in which this would get us in trouble?

**Reply by kyranjamie** [reply]
https://github.com/blockstack/stacks-wallet-web/issues/1715 would be helpful in this instance. Our nonce calculation could be modified to count submitted transactions, too.

My objections relate to the fact that I think it's being oversold as a solution to the issue which is entirely caused by by the API data being wrong—the real problem. 

Whenever we use local/cached data, the wallet is guessing/making assumptions about the state of the chain. It brings with it a whole set of complexities, such as to how & when we remove stale txs, user confusion about why their txs aren't on the explorer. Further, it adds a lot more logic to an implementation that we're hoping to change. 

If we go down this route as a solution to the API problems, we should keep the implementation to the absolute bare minimum. If the API is fixed quickly, we should postpone adding this feature altogether.

**Reply by markmhendrickson** [reply]
I agree this should be viewed as a backup mechanism of sorts to smooth over API problems when they occur (and I expect they _will_ recur in the future, even if fixed quickly this week). 

This mechanism is no substitute for resolving the API issues. That needs to happen clearly, regardless of what we do on the client-side.

Keeping it to the bare minimum also sounds prudent to me 👍

**Reply by beguene** [reply]
I agree with @kyranjamie. The issue is that we have no idea if the api is returning the correct values. If the data returned by the API is not accurate the wallet has no clue so we can't switch to the backup nonce calculation. If we have a way to know that there is a diff between the API production follower and the extra follower that would help the wallet know that something is wrong.

**Reply by markmhendrickson** [reply]
I'm confused since @kyranjamie affirmed `#1715 would be helpful in this instance. Our nonce calculation could be modified to count submitted transactions, too.`. Also this related comment: https://github.com/blockstack/stacks-wallet-web/discussions/1712#discussioncomment-1356147

@beguene @kyranjamie are your senses that yes, #1715 would be helpful though imperfect (since client-side caching and nonce calculation could get out of whack). Or no, it's not helpful at all because `the wallet has no clue so we can't switch to the backup nonce calculation`?

What is the "extra follower" in mind here, another API instance referenced by the wallet to cross compare data? I presume we're not adding more API instance dependencies here.

**Reply by yknl** [reply]
Chiming in as another wallet developer. The situation with the API is not that it isn't showing the transactions, but it is showing conflicting state information. The pending transactions and the nonce returned by the API don't match up. There's nothing the wallet client can do to generate the correct state in this case.

**Reply by markmhendrickson** [reply]
Here's the current scenario I have in mind essentially:

1. API is experiencing current propagation issues in which mempool-based / pending transactions aren't shown reliably, but ones in anchor blocks are shown reliably.
2. User's wallet has no pending transactions and nonce `5`, which is accurately indicated by the API
3. User broadcasts transaction (tx1) with nonce `5` but the wallet doesn't show anything to suggest that it "went through". This transaction ultimately confirms.
4. User immediately broadcasts same transaction again (tx2), also with nonce `5` since the wallet hasn't updated its calculation, with or without the API's help. This transaction ultimately fails.
5. User immediately broadcasts a different transaction (tx3), also with nonce `5`. This transaction also ultimately fails

With this caching patch:

1. API is experiencing current propagation issues in which mempool-based / pending transactions aren't shown reliably, but ones in anchor blocks are shown reliably.
2. User's wallet has no pending transactions and nonce `5`, which is accurately indicated by the API
3. User broadcasts transaction (tx1) with nonce `5` but the wallet recognizes the API isn't recognizing it. So it switches to nonce calculation backup mode for future transactions, until all submitted transactions _are_ recognized. This transaction is shown immediately in the UI as a client-side cached value, and the transaction ultimately confirms on the network (perhaps going straight from "submitted" to "confirmed" state)
4. User doesn't broadcast same transaction again, because they see the first went through.
5. User immediately broadcasts a different transaction (tx2), which the wallet gives nonce `6`. This transaction is shown immediately and ultimately confirms.

Is that the same scenario you have in mind, or are we thinking of different interactions?

**Reply by beguene** [reply]
#1715 will correctly increase the nonce based on local activity but the api will take precedence because the api endpoint we use to get the latest correct nonce WILL return a value but incorrect. The wallet does not know that this value is incorrect.

**Reply by beguene** [reply]
@markmhx Regarding the extra follower ref https://github.com/hirosystems/devops/issues/797#issuecomment-920311268
If I understood correctly I think some nodes (extra followers) which are not api production node are picking up the tx but the prod node does not know about it.

**Reply by markmhendrickson** [reply]
> The wallet does not know that this value is incorrect.

Couldn't the wallet use its local nonce value if it has any submitted transactions that haven't been confirmed as pending or "better" by the API? That seems like a reliable heuristic to me that "API is having problems propagating, resort to backup mechanism", unless I'm missing something?

**Reply by fbwoolf** [reply]
Trying to catch up on all the discussion here, but I agree with @kyranjamie and @beguene on this. I believe trying to handle this client-side will likely create more issues for us.

**Reply by aulneau** [reply]
> Couldn't the wallet use its local nonce value if it has any submitted transactions that haven't been confirmed as pending or "better" by the API? That seems like a reliable heuristic to me that "API is having problems propagating, resort to backup mechanism", unless I'm missing something?

This is my thinking too, and I believe it to be one that can work.

**Comment by kyranjamie** [comment]
While the API isn't working properly, we should to communicate to our users that the API is experiencing problems.

This could be achieved by implementing a wallet instability message, similarly to how is described in https://github.com/blockstack/stacks-wallet-web/discussions/1411

**Reply by andresgalante** [reply]
Big plus one for this, more than anything else clear communication is key at times of uncertainty.

**Reply by kyranjamie** [reply]
Could we get this spec'd out @markmhx @jasperjansz? We should consider there being different variants, more than 1 at a given time, link to external resource, maximum length, whether it's dismissible etc.

**Reply by fbwoolf** [reply]
Plus one for this too.
